### PR TITLE
Update ocaml-dockerfile (BuildKit 1.4, wrapped library, Windows fixes)

### PR DESCRIPTION
**Pull request by MisterDA**
This PR changes the generated Dockerfiles on Linux to use the BuildKit syntax, to support here-documents and the `COPY --link` flag to allow more efficient copy. The BuildKit syntax isn't supported on Windows.

Requires:
- <https://github.com/avsm/ocaml-dockerfile/pull/99>.

### CODE REVIEW

File: src/pipeline.ml [hunk 1]
```diff
@@ -88,8 +91,13 @@ let install_compiler_df ~distro ~arch ~switch ?windows_port opam_image =
   run "opam install -y %s" depext @@
   maybe_install_secondary_compiler run os_family switch @@
   entrypoint_exec (Option.to_list personality @ opam_exec) @@
-  (match os_family with `Linux | `Cygwin -> cmd "bash" | `Windows -> cmd_exec ["cmd.exe"]) @@
-  copy ~src:["Dockerfile"] ~dst:"/Dockerfile.ocaml" ()
+  (match os_family with
+   | `Linux | `Cygwin ->
+     cmd "bash" @@
+     copy ~link:true ~src:["Dockerfile"] ~dst:"/Dockerfile.ocaml" ()
```
Comment (maiste) [inline src/pipeline.ml:97]: Does `~link` only work on Linux?

File: src/pipeline.ml [hunk 2]
```diff
@@ -74,6 +74,9 @@ let install_compiler_df ~distro ~arch ~switch ?windows_port opam_image =
     let additional_packages = Ocaml_version.Opam.V2.additional_packages switch in
     String.concat "," (Printf.sprintf "%s.%s" package_name package_version :: additional_packages)
   in
+  (match os_family with
+   | `Linux -> parser_directive (`Syntax "docker/dockerfile:1.4")
```
Comment (maiste) [inline src/pipeline.ml:78]: Does it mean that we don't support the other syntax anymore? Maybe it's a dummy question but will the Dockerfile be able to be executed with an older docker?

File: src/pipeline.ml (hunk 1 above)
Comment (MisterDA) [reply src/pipeline.ml:97]: Yes. I'll remove `` `Cygwin``.

File: src/pipeline.ml (hunk 2 above)
Comment (MisterDA) [reply src/pipeline.ml:78]: The new syntax is an extension, the "old" syntax is still there. This new syntax is opt-in though, and indeed it does require a Docker with BuildKit supported (at least 18.09, released on 2018-11). Support should be ok.

File: src/pipeline.ml [hunk 3]
```diff
@@ -129,25 +142,26 @@ module Make (OCurrent : S.OCURRENT) = struct
             opam @@
             begin match os_family with
             | `Cygwin | `Linux ->
-              copy ~chown:"opam:opam" ~src:["."] ~dst:"/home/opam/opam-repository" () @@
+              copy ~link:true ~chown:"opam:opam" ~src:["."] ~dst:"/home/opam/opam-repository" () @@
               run "opam-sandbox-disable" @@
               run "opam init -k local -a /home/opam/opam-repository --bare" @@
-              run "rm -rf .opam/repo/default/.git"
+              run "rm -rf .opam/repo/default/.git" @@
+              copy ~link:true ~src:["Dockerfile"] ~dst:"/Dockerfile.opam" ()
```
Comment (maiste) [inline src/pipeline.ml:149]: It's just to be sure to understand: why do you need this file whereas before it wasn't required? And, why the Dockerfile becomes a `.opam` file?

**Review by maiste — COMMENTED** [review]
Apart from `Cygwin` LGTM :+1:

File: src/pipeline.ml (hunk 1 above)
Comment (MisterDA) [reply src/pipeline.ml:97]: It's not about symlinks, it's about BuildKit `COPY --link` operation, optimizing the case where the source files don't depend on a previous build stage. This is only supported on Docker for Linux.

**Pull request merged**